### PR TITLE
Fix meridian offset

### LINKED ISSUE #624: Transform/ boundary does not correspond with extent from data
### HydroMT version checks

- [X] I have checked that this issue has not already been reported.
- [X] I have checked that this bug exists on the latest version of HydroMT.

### Reproducible Example

Use: `GridModel` (no underscore) branch from hydromt_wflow

Run this script in debug mode.
n:\Deltabox\Postbox\Dalmijn, Brendan\PS\setup_wflow_test.py
(maybe drag it to a local place with the accompanying geojson).

Break here:
- `setup_laimaps` (wflow.py; hydromt_wflow): https://github.com/Deltares/hydromt_wflow/blob/af63b5a62d3d1ede84232fd9b9ab003477032948/hydromt_wflow/wflow.py#L919

This then goes through `get_rasterdataset` (data_catalog.py) and `get_data` (rasterdataset.py) and `_slice_data` (rasterdataset.py).

Break here:
- `_slice_spatial_dimensions` (rasterdataset.py): https://github.com/Deltares/hydromt/blob/f3be55fb3aaf44e1aceec0fe9acac82d6515768a/hydromt/data_adapter/rasterdataset.py#L547
- `clip_bbox` (raster.py): https://github.com/Deltares/hydromt/blob/f3be55fb3aaf44e1aceec0fe9acac82d6515768a/hydromt/raster.py#L1226

Then in `clip` the actual clipping happens but it need the indices from `clip_bbox`.  

### Current behaviour

While building a model with hydromt_wflow I noticed that the 'LAI' (leaf area index) was full of nodata values. 

When going through the steps (mentioned above) I noticed that clip_bbox in and on itself uses indices (derived from the transform) instead of coordinates (like it did in 0.8).

This example concerns itself with 'modis_lai' (rasterdataset from 'deltares_data'). 
This dataset had an extent in the x direction from -180 to 180. However the transform and the bounds reflect an extent from -180 to 167.99058.

This causes the indexing to be off by something like 2 degrees. 
Which coupled with the `reproject_like` to the overal dataset (with correct coordinates) causes nodata values.

(EDIT)
It happens here:
https://github.com/Deltares/hydromt/blob/f3be55fb3aaf44e1aceec0fe9acac82d6515768a/hydromt/gis_utils.py#L306-L322

'modis_lai' slightly excedes 180 degrees in the east. This prompts this method to kick into action, which then assumes that the data must be from 0-360 degrees (which its not). The eventual `sortby` (return statement) method then solidifies the wrong extent .

### Desired behaviour

I guess that the transform should be correct in regards to the coordinates shown by the data.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by DirkEilander**
## Issue addressed
Fixes #624 

## Explanation
The error was because the the meridian_offset function created an irregular grid as some cells where incorrectly shifted.
The solution checks if the grid is global (x spans 360 degrees) before shifting which is required to keep a regular grid.
In addition we check if the grid is regular when computing the resolution to avoid these funny downstream errors.

## Checklist
- [x] Updated tests or added new tests
- [x] Branch is up to date with `main`
- [x] Tests & pre-commit hooks pass
- [x] Updated documentation if needed
- [x] Updated changelog.rst if needed

## Additional Notes (optional)
Add any additional notes or information that may be helpful.

### CODE REVIEW

**Comment by savente93** [comment]
I'll have a look in just a sec, but also for my info, does this need to be included in v0.9.1?

**Comment by DirkEilander** [comment]
> I'll have a look in just a sec, but also for my info, does this need to be included in v0.9.1?

Yes, see also the updated #647

File: hydromt/gis_utils.py [hunk 1]
```diff
@@ -303,17 +303,22 @@ def axes_attrs(crs):
...
     """re-arrange data along x dim."""
-    if ds.raster.crs is None or ds.raster.crs.is_projected:
-        raise ValueError("The method is only applicable to geographic CRS")
+    w, _, e, _ = ds.raster.bounds
+    if (
+        ds.raster.crs is None
+        or ds.raster.crs.is_projected
+        or not np.isclose(e - w, 360)
+    ):
+        raise ValueError("The method is only applicable to global geographic CRS")
```
Comment (savente93) [inline hydromt/gis_utils.py:314]: would be nice to mention in the error which f the conditions failed, but this is also acceptable

File: hydromt/raster.py [hunk 2]
```diff
@@ -560,18 +555,25 @@ def res(self) -> tuple[float, float]:
...
-                dy = math.hypot(ddx0, ddy0)
+                dxs, dys = np.diff(xs[0, :]), np.diff(ys[:, 0])
+                ddx0 = xs[-1, 0] - xs[0, 0]
+                ddy0 = ys[-1, 0] - ys[0, 0]
+                ddx1 = xs[0, -1] - xs[0, 0]
+                ddy1 = ys[0, -1] - ys[0, 0]
+                dx = math.hypot(ddx1, ddy1) / (xs.shape[1] - 1)  # always positive!
+                dy = math.hypot(ddx0, ddy0) / (xs.shape[0] - 1)
+            # check if regular grid; not water tight for rotated grids, but close enough
+            # allow rtol 0.05% * resolution for rounding errors in geographic coords
```
Comment (savente93) [inline hydromt/raster.py:572]: should we document this tolerance somewhere? Also the tolerance below is set using `atol` not `rtol`, is this intentional?

File: hydromt/raster.py [hunk 3]
```diff
@@ -2364,8 +2366,15 @@ def to_xyz_tiles(
...
                 if any((s > 0 for s in pad_sizes)):
-                    pad_sizes = {d: (0, s) for d, s in zip(dims, pad_sizes) if s > 0}
-                    temp = temp.pad(pad_sizes, mode="constant", constant_values=nodata)
+                    pad_dict = {d: (0, s) for d, s in zip(dims, pad_sizes) if s > 0}
+                    temp = temp.pad(pad_dict, mode="constant", constant_values=nodata)
+                    # pad nan coordinates; this is not done by xarray
+                    for d, s in zip(dims, pad_sizes):
+                        if s > 0:
+                            coords = temp[d].values
+                            res = dst_res if d == x_dim else -dst_res
```
Comment (savente93) [inline hydromt/raster.py:2375]: can resolutions be negative? this seems odd to me, but perhaps I'm not quite understanding what's happening here.

**Review by savente93 — COMMENTED** [review]
couple of small comments, but mostly looks fine. Once the tests pass and Brendan has confirmed it solves his problem I'll be happy to merge

File: hydromt/raster.py (hunk 2 above)
Comment (DirkEilander) [reply hydromt/raster.py:572]: This is intentional to work with resolution in projected units (metre) and geographic units degrees

File: hydromt/raster.py (hunk 3 above)
Comment (DirkEilander) [reply hydromt/raster.py:2375]: Yes, the orientation of the grid is also captured in the res. A grid which is is oriented north > south has a negative y res

**Review by dalmijn — APPROVED** [review]
lgtm and tested against own case.

**Pull request merged**